### ISSUE TITLE
Move the database to separate folder to make it easier to create a docker image

### ISSUE THREAD

**Issue by mp-se** [thorrak/fermentrack#511]
I've been building a docker image for fermentrack but docker seams to have issues with importing single files into the container. 

In my dockerfile I move the database file to a subdirectory called db and then import this volume. Since I've done it this way the docker image would stop working if git is updating to a new version so I'm proposing a change that moves the db.sqlite3 file to it's own subfolder and still be able to do git upgrades. 

My other option is to build images per fermentrack release and disable the git integration which i'm testing now. 

I have the docker build running here mp-se/fermentrack-docker and under docker hub mpse2/fermentrack-docker. 

let me know your thoughts on this approach.....

**Comment by thorrak** [comment]
For the past few months, I've been working on a new project designed to be run as a Docker container primarily. This project replaces SQLite with a Postgres container. Whenever I manage to get Fermentrack moved over to have native docker support, I plan to do the same. 

That said, there's nothing I can think of that would prevent going down this path now.

**Comment by mp-se** [comment]
Ok, then I will start working on this change if thats ok with you. Don't see that it should be to complicated. The db file is not references in to many places. I found the following places;

fermentrack_django/settings.py
app/management/commands/fix_sqlite_for_django_2.py

The following will need to be updated to secure that the db file is in the sub-directory (or moved to the correct place) in case it's a git upgrade;
utils/upgrade3.sh
utils/force_upgrade3.sh

A fresh install would not be an issue since the db file does not exist yet.

**Comment by thorrak** [comment]
`fix_sqlite_for_django_2` can be ignored as it's specifically to resolve a SQLite error (Postgres would have been fine) and only applies to upgrades - not new installs. 

In general, how are you handling the GIT updating inside the Docker container?

**Comment by lbussy** [comment]
Mind you I have not looked at doing this so I could be full of it:  Wouldn't it be easier to use a `.gitignore` for the database?

**Comment by mp-se** [comment]
Currently I have just disabled the git update process since updating the image would be defying the purpose of having a docker image... The update process should be more in the line of updating the image... so some kind of lookup to check if there is a newer image would be needed in to replace the git update process.. I have not yet figured that part out yet... Should be possible to check versions in docker hub or on a webpage... I would expect that any git upgrade/write to the image would be temporary during the runtime and it will revert to the initial state when started/restarted. 

The build process is quite simple and it was really easy to build a new image when the repository is changed (just a few clicks in docker hub). Check-in to master becomes image:latest and anything with a version tag attached will be an release. So whenever I do a change to my fermentrack-docker repo a new image is built and published. 

I have the setup in a separate repository for now (mp-se/fermentrack-docker) but it would be quite easy to make it part of the main repository and integrate the build process docker hub.

 I just want to make sure the image is stable enough to work. I have it running om my synology NAS at the moment which seams to work fine. I have noticed that the brew-pi scripts seams to crash but the rest seams to work fine.

I have not tested the firmware update process via usb yet, but I have other containers running with that function so I just need to get the time to test that.

**Comment by vanosg** [comment]
Concur on that- you want code updates to occur from outside the image itself, via a new 'docker pull' to get the latest code/image.

**Comment by mp-se** [comment]
> `fix_sqlite_for_django_2` can be ignored as it's specifically to resolve a SQLite error (Postgres would have been fine) and only applies to upgrades - not new installs.
> 
> In general, how are you handling the GIT updating inside the Docker container?

But there could probably be a case that this needs to be run if the user has not upgrade in a long time... If we do a git upgrade where the db file is moved this script would fail.

**Comment by vanosg** [comment]
Honestly? As part of the Dockerfile, run a sed command or similar that disables the update button in the UI after it is downloaded via git, and changes the label to 'update via docker' 

You can also check if the db file is present as part of the entrypoint script used when starting up. I'm not sure what it does or what you're trying to do (only skimmed the thread), but if the db file is there, do X, if not, move it and do Y (if that is what you're trying to do, again, apologies if no)

**Comment by mp-se** [comment]
> Mind you I have not looked at doing this so I could be full of it: Wouldn't it be easier to use a `.gitignore` for the database?

This will not really help. We need to separate the code (binaries) from the data which is stored outside the image. Docker will mount external catalogue (volumes) into the container for the things that should be persistent. This way you can just replace the image and upgrade the system. 

The issue is that the db.sqlite3 is stored in the /fermentrack library and docker is not really good at handling single files that is mounted into the container. And having the whole installation of fermetrack outside the container is little use.... This way we cannot ship a prebuilt/tested image. 

If the database file does not exist (as per a new installation) it will create a directory with that filename and the database creation will fail. So in my docker process I have moved the database to a subdirectory in order for this to work... but that means that I cannot push that change to the main fermentrack repo unless we move the database as part of the standard installation.. Hope this helps to explain things.

**Comment by thorrak** [comment]
> Concur on that- you want code updates to occur from outside the image itself, via a new 'docker pull' to get the latest code/image.

Why? Assuming the container persists across restarts, what does it matter if you update from git within the container vs. updating the container from outside? 

> Honestly? As part of the Dockerfile, run a sed command or similar that disables the update button in the UI after it is downloaded via git, and changes the label to 'update via docker'

Ceteris paribus I agree that updating the docker image is the better way to go, but unless there's something I don't know there's no way for a docker image to "break the glass" and trigger an update to itself. While users currently can trigger most updates through a button in the UI, for docker deployments users would have to log into their Pi and trigger each update manually if the git updates were disabled.

**Comment by lbussy** [comment]
Got it - I have not tried to do what you are doing.

Maybe I'm the slow kid today (shut it, John) but what's the value in having to use Git for deployment only to make a container with it?  Seems to me the value in Docer is to deploy the container without such machinations?

**Comment by thorrak** [comment]
> If the database file does not exist (as per a new installation) it will create a directory with that filename and the database creation will fail. So in my docker process I have moved the database to a subdirectory in order for this to work... but that means that I cannot push that change to the main fermentrack repo unless we move the database as part of the standard installation.. Hope this helps to explain things.

Are you using docker compose to generate everything? If so, then there's another way we could do all this (although it would still require (backwards compatible) changes to `settings.py`)

**Comment by lbussy** [comment]
> Ceteris paribus

Quoted, before he removes it. :P

**Comment by mp-se** [comment]
> Honestly? As part of the Dockerfile, run a sed command or similar that disables the update button in the UI after it is downloaded via git, and changes the label to 'update via docker'
> 
> You can also check if the db file is present as part of the entrypoint script used when starting up. I'm not sure what it does or what you're trying to do (only skimmed the thread), but if the db file is there, do X, if not, move it and do Y (if that is what you're trying to do, again, apologies if no)

Its a few more changes to the code, so a simple sed will not really fix everything but I have implemented this part in my repo.

But I like your approach, I will need to change he db path in the django settings file... In my currently build process I have added code that check if there is an environment variable called DOCKER which disable the git integration. I would be quite easy to us this approach for the django settings file as well so the change only is valid when running in a container. 

Thanks for the suggestion, I was making it way to complex..... I just need to figure out a good way to inform the user that there is a newer docker image... but I might be able to use the current git integration to give me that trigger...

**Comment by vanosg** [comment]
> Maybe I'm the slow kid today (shut it, John) but what's the value in having to use Git for deployment only to make a container with it? Seems to me the value in Docer is to deploy the container without such machinations?


When the image is built, it pulls the git repo, and that image becomes 'locked' at the git code it downloaded at the time of build. The image is what is downloaded and distributed. In a docker image, you would "abandon" the current method of updating via git, you would instead just pull a new image. Each container is ephemeral, the container can be stopped and destroyed, then rebuilt and replaced with an absolute minimum set up and configuration to get you back to where you were before- all the 'custom' changes are stored outside the container.

And really, @lbussy, you touched on the difference of approach. @thorrak yes, you would have to log in to the Pi to update the container, but that kind of is the point of Docker- that's just how you do it. Otherwise, there's no point of using it and @mp-se shouldn't waste his time doing this

**Comment by mp-se** [comment]
> > If the database file does not exist (as per a new installation) it will create a directory with that filename and the database creation will fail. So in my docker process I have moved the database to a subdirectory in order for this to work... but that means that I cannot push that change to the main fermentrack repo unless we move the database as part of the standard installation.. Hope this helps to explain things.
> 
> Are you using docker compose to generate everything? If so, then there's another way we could do all this (although it would still require (backwards compatible) changes to `settings.py`)

I mainly put everything into a Dockerfile (i only use docker compose for my dev environment). I dont really have that much experience in docker (i have used it but not really created one before).

**Comment by thorrak** [comment]
> Maybe I'm the slow kid today (shut it, John) but what's the value in having to use Git for deployment only to make a container with it? Seems to me the value in Docker is to deploy the container without such machinations?

In my head, at the moment there are two types of updates that Fermentrack tends to need:

* Code updates (fresh pulls from GitHub)
* Environmental updates (things that would require the user to SSH in for whatever reason)

There's a fair number of Fermentrack's users who aren't necessarily comfortable with Linux/the command line, so minimizing the number of updates that require SSH'ing in is (in my head) one of Fermentrack's core features. 

That said, it's the environmental issues/updates that cause the most problems (see #463). If I could replace all the environmental update shenanigans with docker container rebuilds, I would be an _exceptionally_ happy camper. If I could do so while also maintaining the "push button, get `git` updates" interface for the rest of the small changes - even better.

**Comment by thorrak** [comment]
> > Are you using docker compose to generate everything? If so, then there's another way we could do all this (although it would still require (backwards compatible) changes to `settings.py`)
> 
> I mainly put everything into a Dockerfile (i only use docker compose for my dev environment). I dont really have that much experience in docker (i have used it but not really created one before).

Got it. I'm going to go pick some beer up now from a brewery down the street that will ~fuel~ ruin tonight's development efforts -- let me take a stab at making some Docker-focused changes to Fermentrack when I get back.

**Comment by mp-se** [comment]
> > Maybe I'm the slow kid today (shut it, John) but what's the value in having to use Git for deployment only to make a container with it? Seems to me the value in Docker is to deploy the container without such machinations?
> 
> In my head, at the moment there are two types of updates that Fermentrack tends to need:
> 
> * Code updates (fresh pulls from GitHub)
> * Environmental updates (things that would require the user to SSH in for whatever reason)
> 
> There's a fair number of Fermentrack's users who aren't necessarily comfortable with Linux/the command line, so minimizing the number of updates that require SSH'ing in is (in my head) one of Fermentrack's core features.
> 
> That said, it's the environmental issues/updates that cause the most problems (see #463). If I could replace all the environmental update shenanigans with docker container rebuilds, I would be an _exceptionally_ happy camper. If I could do so while also maintaining the "push button, get `git` updates" interface for the rest of the small changes - even better.

I totally agree with you to have a simple approach to that is not more complicated than installation on a raspberry pi. There should be no need for a user to login to the container and do stuff. Just download the container, configure volume/ports and then just go. In the basic set we would need db and data (unless we put the db in the data folder?) Im just concerned that this would compromise security if files from the data directory is accessible via the web server. I have also exported the log directory which is good for troubleshooting but not need for most users. 

If you agree I would propose to do this in two steps,

First PR;
* Add the code changes that makes it possible to build a container and disable git upgrades. 
* Add "something" that can inform a user that they need to download a new image since a newer version exist.. My bet is on trying to reuse the current git integration. 
* For starters I would propose we only build images from the master branch and not from other development branches. 

Second PR:
Once this is works and I can test everything and submit a Dockerfile to the main repo so the whole process can be automated once you push changes to the master branch.

**Comment by vanosg** [comment]
@mp-se  take a look at the develop branch of https://github.com/eggheads/eggdrop-docker for some ideas on how to set up the dockerfile, let me know if you have questions. That is accepted as an official image for the docker library, so it has (mostly) good practices in there you can steal from

**Comment by vanosg** [comment]
Sorry, to be specfic, the develop branch/tag, that's the one that deals with git

**Comment by mp-se** [comment]
> Sorry, to be specfic, the develop branch/tag, that's the one that deals with git

Thanks, it looks like a good start. Lots of stuff to reuse. Will take a stab at it tomorrow

**Comment by thorrak** [comment]
> > There's a fair number of Fermentrack's users who aren't necessarily comfortable with Linux/the command line, so minimizing the number of updates that require SSH'ing in is (in my head) one of Fermentrack's core features.
> If you agree I would propose to do this in two steps,
> 
> First PR;
> 
> * Add the code changes that makes it possible to build a container and **disable git upgrades**.

Just to be clear: I _don't_ support _this_ change. I'm a fan of using Docker to control the environment, but the user needs to maintain the ability to also update via the (current) git workflow as the vast majority of upgrades don't require environmental change.

I do envision that longer-term the upgrade workflow will need to be revised to differentiate between the two types of upgrades and prompt the user to upgrade the docker container as needed, but I'd rather maintain the in-app upgrade workflow even if it's not a best practice with docker.

**Comment by vanosg** [comment]
I don't think this is an either/or situation. The docker image could/should be a parallel release, not an in-lieu-of release.
![image](https://user-images.githubusercontent.com/2440369/97819103-96664180-1c74-11eb-8d0a-db6d67715070.png)
 The people that like and use docker use the docker image, those that don't use what is provided now.

While not impossible, doing a codebase update from inside the container is not ideal and not how a docker container is intended to be used. Not to say it couldn't be used that way, but docker isn't built to support that.

BUT - as a solution, it could be possible to mount a volume to ~/fermentrack or wherever the code is located, and write updates to that dir. That way, the image could be blown away and rebuilt, but the codebase isn't taken with it. However, you still encounter the same issue as earlier- if you're trying to avoid users needing to update docker images, now you still have to when the environment baseline changes. So at some point, you're going to need to do it if you release a docker baseline

But the key point to hammer home here- once it is written, it is fairly easy to support both versions in parallel, and the user can choose which experience they prefer. You're just stuck figuring out how to manage the environment ;)

**Comment by uSlackr** [comment]
What if I have custom code I merge into FT after an upgrade?  Can I change the code inside docker?

\\Greg

**Comment by thorrak** [comment]
> While not impossible, doing a codebase update from inside the container is not ideal and not how a docker container is intended to be used. Not to say it couldn't be used that way, but docker isn't built to support that.
> 
> BUT - as a solution, it could be possible to mount a volume to ~/fermentrack or wherever the code is located, and write updates to that dir. That way, the image could be blown away and rebuilt, but the codebase isn't taken with it. However, you still encounter the same issue as earlier- if you're trying to avoid users needing to update docker images, now you still have to when the environment baseline changes. So at some point, you're going to need to do it if you release a docker baseline
> 
> But the key point to hammer home here- once it is written, it is fairly easy to support both versions in parallel, and the user can choose which experience they prefer. **You're just stuck figuring out how to manage the environment** ;)

That's why I want it to (eventually) _not_ be either/or. I want people using Docker. ;) Call it trauma from the recent Django 3.0 migration, call it the constant issues with environmental problems (I think I have two? three? different "debug environment" functions in Fermentrack at the moment?), call it my obsession with fancy new things - I like the idea of being able to dictate the environment Fermentrack runs in. 

I'll always support non-docker installs for people that want to go that route, but I would eventually like all "single command" installs to end up inside a container.

Although I know it's not how Docker is intended to be used, what I don't understand are the actual negative implications of updating the code inside the container so long as the container itself is persisted. It seems like so long as the container itself is relaunched on reboot, it should work mostly as expected.

**Comment by thorrak** [comment]
> What if I have custom code I merge into FT after an upgrade? Can I change the code inside docker? \\Greg 

Yes, but not easily. The "old" method of installing Fermentrack will still be available, I'll just groan at you if you have issues that are caused by the environment. ;)

**Comment by thorrak** [comment]
A start to this has been pushed to the `docker` branch. I'm working on the fermentrack-tools changes now to handle the actual install.

**Comment by mp-se** [comment]
Mounting the whole fermentrack directory outside the container is not a good idea. Its quite easy that there will be inconsistencies with the normal runtime, python install, python env and the nginx install. 

The best approach is to separate code and data so the image is iself sustained so that the image can be tested and supported as a whole. The optimal approach would be to have separate containers for db, nginx and fermentrack but I would personally not go that route for this small application (and I dont have docker-compose support on my NAS) so multiple containers would be lots of manual work for me. 

Regarding writing to a container that will work (a writable layer is created per container) but it will only be persisted in that instance so it's possible that if this is restarted on a new host all changes will be lost . So in this case the user might be prompted for GIT upgrades when that occurs since this is not part of the base image. However in this case it will probably not be a case that will occur that often since many of us are not running docker in a kubernetes cluster. 

@thorrak let me know if you want any support / help on this. I saw that you got started with the additions including moving to postgres. I will keep my current docker build maintained for now and once this is done I will drop that for the main stream process. 

So yes it would be possible to run standard git upgrades with the container but it's not really the best practice. However we still need to move the files that should be persisted outside the container /data + db.

**Comment by thorrak** [comment]
Hey @mp-se -- How did you handle Circus? Did you just replace it with supervisord?

**Comment by mp-se** [comment]
> Hey @mp-se -- How did you handle Circus? Did you just replace it with supervisord?

Currently im just using supervisor to start circus, i didnt want to do too many changes to fermentrack. And supervsisor was used in the other examples i found.

I guess that one could get rid of supervisor to make it less complex

I will run my setup for a while to check if its stable.

I will look over the launch script tomorrow to do some better error handling and see if i can simplify it.

**Comment by thorrak** [comment]
It's not relevant to this PR (which I think #515 closes), but I think that you (@mp-se) and I are taking two different approaches to our docker implementations.

Your implementation has Redis, Nginx, and Fermentrack all running inside a single Ubuntu-based container, which yields a very similar experience/environment to a typical Fermentrack installation. The approach I'm taking has Redis, Nginx, and Fermentrack all in their own containers, using the non-Alpine python base image for Fermentrack rather than Ubuntu.

I don't know enough to be able to speak to the exact benefits/drawbacks of each approach, but my initial thought is that your approach might be better for users that need a single container for their environment, while my approach might be better if docker-compose can be used to manage everything. I'm working on the install script to leverage docker-compose for new installs, but I can certainly see the benefit of maintaining both.

**Comment by mp-se** [comment]
I think we are both after the same thing, make sure that there is a stable base for running fermentrack and avoid having to sort out issues in technology stack, ngnix, python, django. Your approach is the more pure one and is following the "best" practies, separating the workloads in containers which is a must if you want to have a scalable installation (for instance in a kubernetes cluster). This is the approach I would choose when creating an enterprise grade application. But that requires a q more complex infrastructure which most home users dont have. 

Im running my installation om my NAS which supports starting one container (no support for docker compose) which makes it more complex to setup an environment since I need to do everything manual. I cannot use an installation that uses docker-compose... 

So for me its much easier to have just one containter to start/stop rather than several. But if you are moving towards other databases than sqlite i would consider splitting the installation but then i might want to have the database serve more applications to simplify backup etc. 

I did some research on which image to use and found one blog post that had tested ubuntu and alpine. The conclusion was that ubuntu was about the same size and was more stable in some cases... and since Im using ubuntu for my dev workstation that was the easy path to start. But i will probably try out other options to see the difference. 

I looked at your code and there is a lot of things that is changed and I guess it will take a while to get all of that tested. If you are ok to have a intermediate docker version available I have narrowed down it to the following lines in settings.py, Incorporating your code changes to the master branch is more work and testing. I'm fine with helping out to maintain an simpler version of a container release and I prefer to base this your repository rather than my fork to secure that all users get the same experience. If you want I can submit a new PR with these minimum changes for the master branch. it's the same environment variables as in your docker branch without the additional python libraries that you included. 

if os.getenv("USE_DOCKER", default="no").lower() == "yes":
    DB_DIR = os.path.join(BASE_DIR, "db")
else:
    DB_DIR = BASE_DIR

DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.sqlite3',
        'NAME': os.path.join(DB_DIR, 'db.sqlite3'),
    }
}

**Comment by thorrak** [comment]
> If you want I can submit a new PR with these minimum changes for the master branch. it's the same environment variables as in your docker branch without the additional python libraries that you included.
> 
> if os.getenv("USE_DOCKER", default="no").lower() == "yes":
> DB_DIR = os.path.join(BASE_DIR, "db")
> else:
> DB_DIR = BASE_DIR
> 
> DATABASES = {
> 'default': {
> 'ENGINE': 'django.db.backends.sqlite3',
> 'NAME': os.path.join(DB_DIR, 'db.sqlite3'),
> }
> }

Sure thing - please submit the PR against `dev` though. I'll do a test install later this afternoon and make sure non-docker installs still work as expected, then merge to `master`.

That said, I'd rather use `True` than `yes` as that would then match [the way the setting will be](https://github.com/thorrak/fermentrack/blob/7c8849e846449b923d623cbe82331caf16ffe72b/.envs/.prod-sample/.django#L8) once the docker branch is merged. `django-environ` will convert that to a python bool, but even just comparing the string "True" should work I think (e.g. `if os.getenv("USE_DOCKER", default="no").lower() == "true":`)

**Comment by vanosg** [comment]
I'm losing track of some of this conversation, but the "right" way to do this would be @thorrak's approach, you generally want to keep one main process per container- a docker-compose file makes this easier to glue together and provide an easier experience for the user. The benefit here is, I think, you can do away with circus now, and let docker manage the processes. Basically, if the process started by the container dies (lets say apache2 for this example), the entire container will stop running along with it. With a docker-compose file (and restart:always set), docker can monitor that container (and by extension, that process) and restart it if it dies.

Putting multiple processes into one container does not allow you to easily reap the benefits of this approach, and you would have to add something more heavyweight like upstart/systemd/circus(?)/supervisor to manage all the processes. 

The former is harder to set up the first time (for the developer creating the docker-compose file, at least), but you get granularity out of it. For instance, if you want to update apache2 to the next newer version, you can blow away just the apache2 container and restart with a new one, without interrupting the operation of the other containers, like database logging.
